source/Higgs/UntruncatedAtMostTwo.lagda

```diff
@@ -0,0 +1,244 @@
+
+almost-constantly-is-constant
+ : {𝓤 𝓥 : Universe}
+ → {X : 𝓤 ̇ } {Y : 𝓥 ̇ }
+ → (X-discrete : is-discrete X)
+ → (x : X)
+ → (y y' : Y)
+ → y ＝ y'
+ → constantly y ＝ almost-constantly X-discrete x y' y
+almost-constantly-is-constant {_} {_} {X} {Y} X-discrete x y _ refl = dfunext fe I
+ where
+  I : (x' : X)
+    → constantly y x' ＝ almost-constantly X-discrete x y y x'
+  -- possibly rewrite this with dep-cases?
+  -- it's more obvious and spartan but much more awkward
+  I x' with X-discrete x' x
+  ... | inl _ = refl
+  ... | inr _ = refl
+
+at-most-discrete-gives-discrete
+  : {𝓤 𝓥 : Universe}
+  → (X : 𝓤 ̇) (Y : 𝓥 ̇)
+  → is-discrete X
+  → ((f : X → Y) → f has-a-repetition)
+  → is-discrete Y
+at-most-discrete-gives-discrete X Y X-discrete f-ph y y' = V VI
+ where
+
+  repeat-indices : (X → Y)
+                 → X × X
+  repeat-indices f =
+    (λ (x , x' , _) → (x , x'))
+    (f-ph f)
+
+  repeat-is-repeat : (f : X → Y)
+                   → let (x , x') = repeat-indices f in
+                   f x ＝ f x'
+  repeat-is-repeat f =
+    let (x , x' , _ , pf) = f-ph f
+    in pf
+
+  repeat-distinct : (f : X → Y)
+                → let (x , x') = repeat-indices f in
+                x ≠ x'
+  repeat-distinct f =
+    let (x , x' , pf , _) = f-ph f
+    in pf
+
+  f₁ = constantly y
+  ix₁ = repeat-indices f₁
+  f₂ = almost-constantly X-discrete (pr₁ ix₁) y' y
+  ix₂ = repeat-indices f₂
+
+  I : y ＝ y' → ix₁ ＝ ix₂
+  I e = ap repeat-indices
+    (almost-constantly-is-constant X-discrete (pr₁ ix₁) y y' e)
+
+  II : (x : X)
+     → (pr₁ ix₁ ≠ x)
+     → y ＝ f₂ x
+  II x ne with X-discrete x (pr₁ ix₁)
+  ... | inl e = 𝟘-elim (ne (e ⁻¹))
+  ... | inr _ = refl
+
+  III : f₂ (pr₁ ix₁) ＝ y'
+  III with X-discrete (pr₁ ix₁) (pr₁ ix₁)
+  ... | inl e = refl
+  ... | inr ne = 𝟘-elim (ne refl)
+
+  IV : ix₁ ＝ ix₂ → y ＝ y'
+  IV e =
+    y            ＝⟨ II (pr₂ ix₁) (repeat-distinct f₁) ⟩
+    f₂ (pr₂ ix₁) ＝⟨ ap (f₂ ∘ pr₂) e ⟩
+    f₂ (pr₂ ix₂) ＝⟨ refl ⟩
+    f₂ (pr₂ (repeat-indices f₂)) ＝⟨ repeat-is-repeat f₂ ⁻¹ ⟩
+    f₂ (pr₁ (repeat-indices f₂)) ＝⟨ refl ⟩
+    f₂ (pr₁ ix₂) ＝⟨ ap (f₂ ∘ pr₁) (e ⁻¹) ⟩
+    f₂ (pr₁ ix₁) ＝⟨ III ⟩
+    y' ∎
+
+  V : is-decidable (ix₁ ＝ ix₂) → is-decidable (y ＝ y')
+  V (inl e) = inl (IV e)
+  V (inr ne) = inr (ne ∘ I)
+
+  VI : is-decidable (ix₁ ＝ ix₂)
+  VI = ×-is-discrete X-discrete X-discrete ix₁ ix₂
+
+\end{code}
+
+We may write the untruncated form of the at-most-2 lemma in this form
+
+\begin{code} 
+
+at-most-two-is-pigeonhole
+  : {𝓤 : Universe}
+  → {X : 𝓤 ̇}
+  → ((x y z : X) → (z ＝ x) + (x ＝ y) + (y ＝ z))
+  → (f : Fin 3 → X)
+  → f has-a-repetition
+at-most-two-is-pigeonhole at-most-2 f = II I
+ where
+  v1 v2 v3 : Fin 3
+  v1 = inr ⋆
+  v2 = inl (inr ⋆)
+  v3 = inl (inl (inr ⋆))
+
+  true-when : Fin 3
+            → Fin 3
+            → 𝓤 ⁺ ̇
+  true-when (inr _) (inr _) = 𝟙
+  true-when (inl (inr _)) (inl (inr _)) = 𝟙
+  true-when (inl (inl (inr _))) (inl (inl (inr _))) = 𝟙
+  {-# CATCHALL #-}
```

Comment:
I am not sure what `{-# CATCHALL #-}` is doing.

Reply:
Please fix this by writing the missing cases. Are you using case-split on emacs (`ctrl-c-ctrl-c`) to generate the cases? Using this, the `{-# CATCHALL #-}` should not be needed. Thanks.

Reply:
Okay, I've done this, it's harder to read now in my opinion.
The issue was just a warning from Agda saying to the effect that the syntax didn't match what would reduce definitionally.

The automatic case-split here was quite tedious since it makes 9 cases and requires splitting each time.